## Code for Project 3 

#### Writing hyperlink matrix from 4.3

\begin{equation}
H = 
\begin{bmatrix} 
0 & 0 & 1/3 & 0 & 0 & 0\\ 
1/2 & 0 & 1/3 & 0 & 0 & 0 \\
1/2 & 1 & 0 & 0 & 0 & 0 \\
0 & 0 & 0 & 0 & 1/2 & 1 \\
0 & 0 & 1/3 & 1/2 & 0 & 0 \\
0 & 0 & 0 & 1/2 & 1/2 & 0
\end{bmatrix}
\end{equation}

#### Iterative Process:

In [1]:
import numpy as np
H = np.matrix([
    [0,   0, 1/3, 0,   0,   0],
    [1/2, 0, 1/3, 0,   0,   0],
    [1/2, 1, 0,   0,   0,   0],
    [0,   0, 0,   0,  1/2,  1],
    [0,   0, 1/3, 1/2, 0,   0],
    [0,   0, 0,   1/2, 1/2, 0]
])
n = H.shape[0]
x = np.matrix([[1/6], [1/6], [1/6], [1/6], [1/6], [1/6]])
iterations = 20 # arbitrary number
for i in range(iterations):
    x = H * x 
print(x)

[[0.00285311]
 [0.004559  ]
 [0.00715766]
 [0.43695276]
 [0.2205959 ]
 [0.32788157]]


#### What must be true about a collection of n pages such that an nxn hyperlink matrix H is a stochastic matrix?

A matrix is stochastic if all entries are nonnegative (greater or equal to zero) and if each column sums to 1. For H to be stochastic, each page must have at least one outgoing link. This is because each column of H represents a page and the entries in that column represent the probability of clicking a link from that page to others (outgoing). If a page had no outgoing links, then it's column would be all zeros. If even one page has no outgoing links, then H won't be stochastic. 

#### Finishing Theorem 4.9

\begin{equation}
x_{equilib} = \lim_{k \to \infty}A^kx_0 = v
\end{equation}
where v is the eigenvector corresponding to eigenvalue $\lambda$ = 1 and it is normalized so its entries sum to 1. For a stochastic matrix, $\lambda$ = 1 is always an eigenvalue. From the given theorem, all other eigenvalues satisfy $|\lambda_j|$ < 1, so $\lambda^k_j$ -> 0. This means that all other components except the $\lambda$ = 1 disappear as k -> $\infty$

#### What does this mean for PageRank?

Instead of having to do the iterative process, we only need to find the eigenvector that corresponds to the eigenvalue of 1, and normalized so that its entries sum to 1, as seen from the theorem above. This turns the problem into a linear algebra one instead of a long, tedious iterative process. 

In [3]:
# using the theorem to find our pagerank vector
w, v = np.linalg.eig(H)
print(w)
print("")
print(v)

[ 1.        +0.00000000e+00j  0.83624392+0.00000000e+00j
 -0.41812196+1.56454205e-01j -0.41812196-1.56454205e-01j
 -0.5       +2.75142433e-09j -0.5       -2.75142433e-09j]

[[ 3.58286687e-16+0.00000000e+00j -2.19697400e-01+0.00000000e+00j
   9.18493350e-02-1.69482811e-01j  9.18493350e-02+1.69482811e-01j
   1.17191869e-16+6.21386714e-24j  1.17191869e-16-6.21386714e-24j]
 [ 5.81610806e-16+0.00000000e+00j -3.51057039e-01+0.00000000e+00j
  -7.10188410e-02-2.77538411e-02j -7.10188410e-02+2.77538411e-02j
   2.19025728e-32+1.47415721e-24j  2.19025728e-32-1.47415721e-24j]
 [ 8.71880917e-16+0.00000000e+00j -5.51161843e-01+0.00000000e+00j
  -3.56637764e-02+2.55704099e-01j -3.56637764e-02-2.55704099e-01j
  -3.54505405e-16-1.07980138e-23j -3.54505405e-16+1.07980138e-23j]
 [ 7.42781353e-01+0.00000000e+00j  5.76885127e-01+0.00000000e+00j
  -5.80145151e-01+5.55076822e-02j -5.80145151e-01-5.55076822e-02j
  -7.07106781e-01+0.00000000e+00j -7.07106781e-01-0.00000000e+00j]
 [ 3.71390676e-01+0.00000000e+0

From this, we can see that the first eigenvalue is 1, which corresponds to the first column of v:

In [4]:
v[:,0]

matrix([[3.58286687e-16+0.j],
        [5.81610806e-16+0.j],
        [8.71880917e-16+0.j],
        [7.42781353e-01+0.j],
        [3.71390676e-01+0.j],
        [5.57086015e-01+0.j]])

In [10]:
# Now we normalize it:
x_eig = np.real(v[:, 0])
x_eig = x_eig / np.sum(x_eig)
print(x_eig)

[[2.14381429e-16]
 [3.48007782e-16]
 [5.21691381e-16]
 [4.44444444e-01]
 [2.22222222e-01]
 [3.33333333e-01]]


This gives us:
\begin{equation}
x = 
\begin{bmatrix} 
2.14 * 10^{-16} \\
3.48* 10^{-16} \\
5.22* 10^{-16} \\
0.4444 \\
0.2222 \\
0.3333
\end{bmatrix}
\end{equation}

So page importance order: 4, 6, 5, 3, 2, 1

These numbers are fairly close to the numbers we got in question 2, with the first three pages not having much importance and the last three holding the majority. The first three pages have pretty different numbers but we can summarize from both of the methods that they are not of importance. The last three pages (4, 5, 6) have very comparable numbers. The second approach results in an answer that is rounded up, but can be considered a good tradeoff with the amount of computational time saved.

#### Figure 4.4

\begin{equation}
H = 
\begin{bmatrix}
0 & 0 & 0 & 0 & 0 & 0 & 1/3 & 0 \\
1/3 & 0 & 1/3 & 0 & 0 & 0 & 0 & 0 \\
1/3 & 1 & 0 & 0 & 1/4 & 0 & 1/3 & 0 \\
0 & 0 & 0 & 0 & 1/4 & 1 & 0 & 1 \\
0 & 0 & 1/3 & 1/3 & 0 & 0 & 0 & 0 \\
0 & 0 & 0 & 1/3 & 1/4 & 0 & 0 & 0 \\
1/3 & 0 & 1/3 & 0 & 1/4 & 0 & 0 & 0 \\
0 & 0 & 0 & 1/3 & 0 & 0 & 1/3 & 0
\end{bmatrix}
\end{equation}

\begin{equation}
x_0 = 
\begin{bmatrix}
1/8 \\
1/8 \\
1/8 \\
1/8 \\
1/8 \\
1/8 \\
1/8 \\
1/8 \\
\end{bmatrix}
\end{equation}

In [7]:
# Using the iterative process:
H_2 = np.matrix([
    [0, 0, 0, 0, 0, 0, 1/3, 0],
    [1/3, 0, 1/3, 0, 0, 0, 0, 0],
    [1/3, 1, 0, 0, 1/4, 0, 1/3, 0],
    [0, 0, 0, 0, 1/4, 1, 0, 1],
    [0, 0, 1/3, 1/3, 0, 0, 0, 0],
    [0, 0, 0, 1/3, 1/4, 0, 0, 0],
    [1/3, 0, 1/3, 0, 1/4, 0, 0, 0],
    [0, 0, 0, 1/3, 0, 0, 1/3, 0]
])
n = H_2.shape[0]
x_2 = np.matrix([[1/8], [1/8], [1/8], [1/8], [1/8], [1/8], [1/8], [1/8]])
iterations = 20 # arbitrary number
for i in range(iterations):
    x_2 = H_2 * x_2
print(x_2)

[[0.02912536]
 [0.05192892]
 [0.12583937]
 [0.29859765]
 [0.14242972]
 [0.13551236]
 [0.08723464]
 [0.129332  ]]


In [8]:
# using eigenvalues
w_2, v_2 = np.linalg.eig(H_2)
print(w_2)
print("")
print(v_2)

[ 1.00000000e+00+0.j         7.54911941e-01+0.j
 -8.14094350e-01+0.j        -7.44156380e-01+0.j
 -2.82267009e-01+0.j         4.28028987e-02+0.2144428j
  4.28028987e-02-0.2144428j  7.91838581e-17+0.j       ]

[[ 6.99885954e-02+0.j          1.17166616e-01+0.j
   1.95592026e-02+0.j         -4.31605071e-02+0.j
   4.31417570e-01+0.j         -4.31010242e-01-0.20036449j
  -4.31010242e-01+0.20036449j -3.86191682e-16+0.j        ]
 [ 1.24424170e-01+0.j          2.84225628e-01+0.j
   7.78746726e-02+0.j          2.63768960e-01+0.j
  -3.23521240e-03+0.j         -3.39724982e-02+0.23570232j
  -3.39724982e-02-0.23570232j  2.80866677e-16+0.j        ]
 [ 3.03283913e-01+0.j          5.26529345e-01+0.j
  -2.09751195e-01+0.j         -5.45695555e-01+0.j
  -4.28677989e-01+0.j          2.75013879e-01+0.20877525j
   2.75013879e-01-0.20877525j  1.78333940e-16+0.j        ]
 [ 7.23215486e-01+0.j         -6.55604382e-01+0.j
  -7.89493213e-01+0.j         -5.66811131e-01+0.j
   8.24855535e-02+0.j         -2.13313303

This also gives us the first column of v as our eigenvector corresponding to eigenvalue 1

In [9]:
v_2[:,0]

matrix([[0.0699886 +0.j],
        [0.12442417+0.j],
        [0.30328391+0.j],
        [0.72321549+0.j],
        [0.34216647+0.j],
        [0.32661345+0.j],
        [0.20996579+0.j],
        [0.31106042+0.j]])

In [11]:
# Now we normalize it:
x_eig2 = np.real(v_2[:, 0])
x_eig2 = x_eig2 / np.sum(x_eig2)
print(x_eig2)

[[0.02903226]
 [0.0516129 ]
 [0.12580645]
 [0.3       ]
 [0.14193548]
 [0.13548387]
 [0.08709677]
 [0.12903226]]


Here are the two vectors we got:
From the iterative process: 

\begin{equation}
x = 
\begin{bmatrix}
0.02912536 \\
0.05192892 \\
0.12583937 \\
0.29859765 \\
0.14242972 \\
0.13551236 \\
0.08723464 \\
0.129332 
\end{bmatrix}
\end{equation}

From using eigenvector:

\begin{equation}
x = 
\begin{bmatrix}
0.02903226 \\
0.0516129 \\
0.12580645 \\
0.3 \\
0.14193548 \\
0.13548387 \\
0.08709677 \\
0.12903226
\end{bmatrix}
\end{equation}

As you can see, the results are very comparable. The page rank numbers are very similar. Both agree that page importance order is: 4, 5, 6, 8, 3, 7, 2, 1